File: cpp/14.longest-common-prefix.cpp
/*
 * @lc app=leetcode.cn id=14 lang=cpp
 *
 * [14] 最长公共前缀
 *
 * https://leetcode-cn.com/problems/longest-common-prefix/description/
 *
 * algorithms
 * Easy (41.05%)
 * Total Accepted:    727.7K
 * Total Submissions: 1.7M
 * Testcase Example:  '["flower","flow","flight"]'
 *
 * 编写一个函数来查找字符串数组中的最长公共前缀。
 * 
 * 如果不存在公共前缀，返回空字符串 ""。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 
 * 输入：strs = ["flower","flow","flight"]
 * 输出："fl"
 * 
 * 
 * 示例 2：
 * 
 * 
 * 输入：strs = ["dog","racecar","car"]
 * 输出：""
 * 解释：输入不存在公共前缀。
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * 1 <= strs.length <= 200
 * 0 <= strs[i].length <= 200
 * strs[i] 仅由小写英文字母组成
 * 
 * 
 */
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string common_prefix;
        
        if (strs.size() == 0){
            return common_prefix;
        }
        
        if (strs.size() == 1){
            return strs[0];
        }
        
        int min_len = 201;
        for (int i=0; i<strs.size(); ++i){
            if (strs[i].size() < min_len){
                min_len = strs[i].size();
            }
        }

        for (int i=0; i<min_len; ++i){
            set<char> s;
            for (int j=0; j<strs.size(); ++j){
                s.insert(strs[j][i]);
            }
            if (s.size() == 1){
                common_prefix += strs[0][i];
            } else {
                return common_prefix;
            }
        }
        
        return common_prefix;

    }
};

File: python/14.longest-common-prefix.py
#
# @lc app=leetcode.cn id=14 lang=python3
#
# [14] 最长公共前缀
#
# https://leetcode-cn.com/problems/longest-common-prefix/description/
#
# algorithms
# Easy (39.09%)
# Total Accepted:    431.8K
# Total Submissions: 1.1M
# Testcase Example:  '["flower","flow","flight"]'
#
# 编写一个函数来查找字符串数组中的最长公共前缀。
# 
# 如果不存在公共前缀，返回空字符串 ""。
# 
# 
# 
# 示例 1：
# 
# 
# 输入：strs = ["flower","flow","flight"]
# 输出："fl"
# 
# 
# 示例 2：
# 
# 
# 输入：strs = ["dog","racecar","car"]
# 输出：""
# 解释：输入不存在公共前缀。
# 
# 
# 
# 提示：
# 
# 
# 0 
# 0 
# strs[i] 仅由小写英文字母组成
# 
# 
#
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        """
        # method 1
        str_len = list(map(len, strs))
        if all(str_len) == 0 or str_len == []:
            return ''
        shortest_len = min(str_len)
        shortest_strs = list(filter(lambda x: len(x) == shortest_len, strs))

        for shortest_str in shortest_strs:
            for i in reversed(range(shortest_len+1)):
                target_str = shortest_str[0: i]
                if all([target_str == s[0: i] for s in strs]):
                    return target_str
        """

        """
        # method 2
        str_len = list(map(len, strs))
        if all(str_len) == 0 or str_len == []:
            return ''
        shortest_len = min(str_len)
        shortest_strs = list(filter(lambda x: len(x) == shortest_len, strs))

        for shortest_str in shortest_strs:
            target_len = shortest_len

            while target_len >= 0:
                len_diff = shortest_len - target_len
                for j in range(len_diff + 1):
                    target_str = shortest_str[j: target_len + j]
                    if all(target_str in single_str for single_str in strs):
                        return target_str

                target_len -= 1
        """
        if len(strs) == 0:
            return ''
        if len(strs) == 1:
            return strs[0]

        min_len = 201
        for s in strs:
            if len(s) < min_len:
                min_len = len(s)

        for i in reversed(range(min_len+1)):
            if len(set([s[0: i] for s in strs])) == 1:
                return strs[0][0: i]
        return ''





